Remove non-standard alias variables

MaggieYingYi opened the pull request:
The global variable alias will cause the wrong variable value since two symbols point two different objects. The aliasee will not change when the aliaser changes. I went through all alias variable and found some alias global variables are only used on Linux or used inside of musl libc library. I have modified the musl libc to remove these alias symbol for the repo target.

This patch will solve the clang lit test failures mentioned in https://github.com/SNSystems/musl-prepo/issues/8#issuecomment-880525308.

paulhuggett commented on ldso/dynlink.c:
```diff
@@ -1742,7 +1742,11 @@ void __dls3(size_t *sp, size_t *auxv)
 
 	/* Find aux vector just past environ[] and use it to initialize
 	 * global data that may be needed before we can make syscalls. */
+#ifndef __REPO__
 	__environ = envp;
+#else
+	environ = envp;
+#endif
```
What's the thinking behind retaining the original (alias-dependant) code?

Keeping both versions means that there is a lot of duplication here. It also seems to me that it would be best to have a single (repo/non-repo) code-base to ensure consistency between the builds.

MaggieYingYi replied:
> 
> 
> What's the thinking behind retaining the original (alias-dependant) code?
> 
>
When I explored the musl libc to answer the question: why musl libc need to use the weak alias, I found the alias global variable could be removed and detailed explanation is given in the https://github.com/SNSystems/musl-prepo/issues/8#issuecomment-880534478.

I want that all other targets keep alias symbol which is the same as upstream musl libc. However, for the repo target, we only support the standard C global variables for both windows and Linux.

As mentioned in the https://github.com/SNSystems/musl-prepo/issues/8#issuecomment-880547716, some alias functions are required. Therefore, I am not sure should we accept this commit (remove some alias global variables) or should we support alias for repo target?
 I am mainly creating this patch for discussing what is the best way to solve the weak_alias macro?

paulhuggett replied:
> I want that all other targets keep alias symbol which is the same as upstream musl libc.

Other targets? We're developing this for the repo only.  As I mentioned, I'm concerned by the duplication that the proposed change introduced.

BTW, it seems to me that quite a lot of the change could be avoided by simply adding #define. e.g.
~~~~C
#define __environ environ
~~~~
Have you tried this?

MaggieYingYi replied:
> 
> BTW, it seems to me that quite a lot of the change could be avoided by simply adding a #define. e.g.
> 
> ```c
> #define __environ environ
> ```
> 
> Have you tried this?
> 

I did consider using `#define` but it doesn't work. Considering change the `__environ.c` as below:

```
diff --git a/src/env/__environ.c b/src/env/__environ.c
index fe8abcf9..2ff09d8e 100644
--- a/src/env/__environ.c
+++ b/src/env/__environ.c
@@ -1,6 +1,8 @@
 #include <unistd.h>

 char **__environ = 0;
-weak_alias(__environ, ___environ);
-weak_alias(__environ, _environ);
-weak_alias(__environ, environ);
+
+#define ___environ __environ
+#define _environ __environ
+#define environ __environ
+
```

When using weak_alias, the symbol table is shown below:
```
__environ.o:     file format elf64-x86-64

SYMBOL TABLE:
0000000000000000 l    df *ABS*  0000000000000000 __environ.c
0000000000000000  w    O .bss.__environ 0000000000000008 ___environ
0000000000000000 g     O .bss.__environ 0000000000000008 __environ
0000000000000000  w    O .bss.__environ 0000000000000008 _environ
0000000000000000  w    O .bss.__environ 0000000000000008 environ
```
When using #define install, the symbol table is shown belown:
```
__environ.o:     file format elf64-x86-64

SYMBOL TABLE:
0000000000000000 l    df *ABS*  0000000000000000 __environ.c
0000000000000000 g     O .bss.__environ 0000000000000008 __environ
```

We could only see `__environ` in the symbol table. Therefore, when we build clang using the musl libc library, we will get the following error:

```
undefined reference to `environ'
``` 

> 
> Required by what? Is it required that they are aliases or simply that the symbols are present?

For example, `weak_alias(euidaccess, eaccess)`, both euidaccess and eaccess refer to the same things which is mentioned in the https://man7.org/linux/man-pages/man3/euidaccess.3.html. If some functions have the same function body but different names (for example exp10 and pow10), I think we might need to support alias.

>
> Does something demand that &fopen equals &fopen64? For example, for a case like this:
> 
> ```c
> weak_alias(fopen, fopen64);
> ```
> 
> Is there a reason why this wouldn’t work (assuming that I have the alias and original the right way round)?
> 
> ```c
> inline FILE *fopen (char const *restrict filename, char const *restrict mode) {
>     return fopen64 (filename, mode);
> }
> ```
>

I think it will work but a lot of musl libc codes need to modify. When we merge the upstream changes, we will meet the merge conflict.

> 
> fopen64() is not a in standard C. Could we simply remove it?
>

I found `fopen()` and `fopen64()` in the https://www.mkssoftware.com/docs/man3/fopen.3.asp. How to make sure the function is not in the standard C?

paulhuggett replied:
>> #define __environ environ
>
> +#define ___environ __environ
> +#define _environ __environ
> +#define environ __environ
> We could only see `__environ` in the symbol table. Therefore, when we build clang using the musl libc library, we will get the following error:
> 
> ```
> undefined reference to `environ'
> ```

You need to define the names as `environ`, not `__environ`.


>  > Required by what? Is it required that they are aliases or simply that the symbols are present?
> 
> For example, `weak_alias(euidaccess, eaccess)`, both euidaccess and eaccess refer to the same things which is mentioned in the https://man7.org/linux/man-pages/man3/euidaccess.3.html. 

This document simply says ”eaccess() is a synonym for euidaccess(), provided for compatibility with some other systems.” This compatibility is almost certainly not necessary for our use-case. If it is, we could add an inline definition of one function that calls the other. 


> > Is there a reason why this wouldn’t work (assuming that I have the alias and original the right way round)?
> > ```c
> > inline FILE *fopen (char const *restrict filename, char const *restrict mode) {
> >     return fopen64 (filename, mode);
> > }
> > ```
> 
> I think it will work but a lot of musl libc codes need to modify. When we merge the upstream changes, we will meet the merge conflict.

Is this a real concern (i.e. do we need to merge upstream changes)? AFAIK, the port is only needed to enable us to link clang for the purposes of an internal deadline.


> > fopen64() is not a in standard C. Could we simply remove it?
> 
> I found `fopen()` and `fopen64()` in the https://www.mkssoftware.com/docs/man3/fopen.3.asp. How to make sure the function is not in the standard C?

For a definitive answer, just check the standard itself (the most recent, free, version is [here](https://web.archive.org/web/20181230041359if_/http://www.open-std.org/jtc1/sc22/wg14/www/abq/c17_updated_proposed_fdis.pdf)). The online [C reference](https://en.cppreference.com/w/c) isn’t authoritative, but it’s convenient and pretty accurate.

MaggieYingYi replied:
>
> You need to define the names as environ, not __environ.
>

Sorry, I am not sure what you mean. Did you mean rename `__environ` to `environ`? (which is the same as this commit, a lot of changes). Could you please modify the `__environ.c` source code to demo what you mean? Many Thanks.

>
> For a definitive answer, just check the standard itself (the most recent, free, version is here). The online C reference isn’t authoritative, but it’s convenient and pretty accurate.
> 

Thanks for the information.

paulhuggett replied:
> > You need to define the names as environ, not __environ.
> 
> Sorry, I am not sure what you mean. 

You wrote:
~~~C
#define ___environ __environ
#define _environ __environ
#define environ __environ
~~~
This needs to be something like:
~~~C
#define ___environ environ
#define __environ environ
#define _environ environ
~~~
You want the name you end up with to be `environ`.

> Could you please modify the __environ.c source code to demo what you mean? 

Not quickly!

MaggieYingYi replied:
I think I got your idea. The original code:

```
char **__environ = 0;
weak_alias(__environ, ___environ);
weak_alias(__environ, _environ);
weak_alias(__environ, environ);
```

If following your idea, I need to change the code to (correct me if I am wrong):
```
char **environ = 0;   // changed from __environ to environ
#define ___environ environ
#define __environ environ
#define _environ environ
```

After this change, we still only have a single symbol `environ` in the symbol table. Therefore, `#define` doesn't work for saving all symbols in the symbol table.

I will remove all other symbols (__environ, ___environ and _environ) and only keep `environ` variable for ELF/Repo etc targets. Is it ok for you?

paulhuggett replied:
> I will remove all other symbols (__environ, ___environ and _environ) and only keep environ variable for ELF/Repo etc targets. Is it ok for you?

Yes. If it’s not in the standard and clang/LLVM doesn't depend on it, we don’t need it!

MaggieYingYi replied:
> 
> 
> > I will remove all other symbols (__environ, ___environ and _environ) and only keep environ variable for ELF/Repo etc targets. Is it ok for you?
> 
> Yes. If it’s not in the standard and clang/LLVM doesn't depend on it, we don’t need it!
>

Thanks Paul. I have updated the patch.

paulhuggett replied:
FWIW, I still see just a single commit.

MaggieYingYi replied:
Yes, it is a singe commit. I removed the previous commit and forced to commit the new change. You need to delete you `remove_weak_alias_variables` branch and checkout again. Sorry about that.

paulhuggett replied:
I’m just looking at the PR ”Files changed” page. How can I know if I’m seeing the right thing? Following our discussion, I was expecting to see #defines being used to limit the amount of code change, but there’s no sign…

MaggieYingYi replied:
As I explained in above comments, #define doesn't work at all. It will not emit the defined symbol in the symbol table. Therefore, there is no added #define.

MaggieYingYi commented:
Thanks for explaining `#define` over the phone. I have removed the previous changes and updated the patch using `#define`.

For the `__eintr_valid_flag` variable, I need to define it as a weak symbol since it is also defined in the sigaction.c file.

Without giving a weak attribute, I will get the following link error:
```
[377/2324] Linking CXX executable bin/llvm-tblgen
FAILED: bin/llvm-tblgen
/usr/bin/ld: /home/maggie/musl/lib/libc_elf.a(__timedwait.t.elf):(.bss.__eintr_valid_flag+0x0): multiple definition of `__eintr_valid_flag'; /home/maggie/musl/lib/libc_elf.a(sigaction.t.elf):(.bss.__eintr_valid_flag+0x0): first defined here
```

paulhuggett commented:
Has the patch been pushed? (Overwriting previous pushes is very confusing to me… please avoid it!)

MaggieYingYi commented:
> Has the patch been pushed? (Overwriting previous pushes is very confusing to me… please avoid it!)


Sorry about the confusion. Yes, it has. You could see https://github.com/SNSystems/musl-prepo/pull/11/files for all file changes.

paulhuggett commented:
That shows me the files, but not that you've committed new work or what has changed since I last reviewed it.

MaggieYingYi commented:
> 
> 
> That shows me the files, but not that you've committed new work or what has changed since I last reviewed it.

Yes, I shouldn't delete the previous commit. Sorry about that.
